src/Client/Connection.cpp

```diff
@@ -404,6 +405,49 @@ void Connection::sendQuery(
     else
         writeStringBinary("" /* empty string is a marker of the end of settings */, *out);
 
+    /// Interserver secret
+    bool interserver_secret = server_revision >= DBMS_MIN_REVISION_WITH_INTERSERVER_SECRET;
+    if (interserver_secret)
+    {
+        /// Cluster
+        ///
+        /// Send cluster name only if there is a cluster secret.
+        /// Since if there is no secret, then no validation should be done.
+        if (!cluster_secret.empty())
+            writeStringBinary(cluster, *out);
+        else
+            writeStringBinary("", *out);
+
+        /// Hash
+        ///
+        /// Send correct hash only for !INITIAL_QUERY, due to:
+        /// - this will avoid extra protocol complexity for simplest cases
+        /// - there is no need in hash for the INITIAL_QUERY anyway
+        ///   (since there is no secure/unsecure changes)
+        if (client_info && !cluster_secret.empty() && client_info->query_kind != ClientInfo::QueryKind::INITIAL_QUERY)
+        {
+#if USE_SSL
+            std::string data(cluster_secret);
+            data += query;
+            data += query_id;
+            data += user;
+            data += client_info->initial_user;
```

Comment:
Since we already get some server info in hello packet, may be server can also send some rand string / salt (ideally it also should be signed with secret) which will be used in hashing to avoid possibility of reusing that hash out of that session.

Reply:
This can't be done in hello packet, since the connection does not bound to specific cluster.

Reply:
> ideally it also should be signed with secret

This can't be done (actually it is also possible, by sending it for each cluster, but this is too much)
Everything else, makes sense

Reply:
Although not sure that it is better then IP address/hostname (todo was left for this for now)

Reply:
Hello - server send random string salt only (no need to be bound to cluster, bounded to connection)

Client then use that salt for creating a signature (i.e. `Sha256(cluster name, query id,..., salt)`). That prevents reusing signature outside of that connection (trafic sniffing attacks etc).

Even safer option is to change that salt for every query but it should be server generated, and require more rounds.

Reply:
Just some more things here - actually that type of 'salt' should be rather called nonce (see https://en.wikipedia.org/wiki/Cryptographic_nonce )

The task we did is called MAC/[HMAC](https://en.wikipedia.org/wiki/HMAC), maybe we can make it a bit more standard way to avoid potential pitfalls (with the security it's better to avoid reinventing things when possible) using standard HMAC-SHA256 (anyway we use openssl, which provide different options for HMAC).

For authentication (with nonce / timestamp / token ) - HMAC is used for example in OAuth: https://tools.ietf.org/html/rfc5849#page-17